// @better-auth/drizzle-adapter/dist/index.mjs
import { createAdapterFactory } from "@better-auth/core/db/adapter";
import { logger } from "@better-auth/core/env";
import { BetterAuthError } from "@better-auth/core/error";
import { and, asc, count, desc, eq, gt, gte, ilike, inArray, isNotNull, isNull, like, lt, lte, ne, notInArray, or, sql } from "drizzle-orm";
//#region src/query-builders.ts
/**
* Case-insensitive LIKE/ILIKE for pattern matching.
* Uses ILIKE on PostgreSQL, LOWER()+LIKE on MySQL/SQLite.
*/
function insensitiveIlike(column, pattern, provider) {
	return provider === "pg" ? ilike(column, pattern) : sql`LOWER(${column}) LIKE LOWER(${pattern})`;
}
/**
* Case-insensitive IN for string arrays.
*/
function insensitiveInArray(column, values) {
	if (values.length === 0) return sql`false`;
	return sql`LOWER(${column}) IN (${sql.join(values.map((v) => sql`LOWER(${v})`), sql`, `)})`;
}
/**
* Case-insensitive NOT IN for string arrays.
*/
function insensitiveNotInArray(column, values) {
	if (values.length === 0) return sql`true`;
	return sql`LOWER(${column}) NOT IN (${sql.join(values.map((v) => sql`LOWER(${v})`), sql`, `)})`;
}
/**
* Case-insensitive equality for strings.
*/
function insensitiveEq(column, value) {
	return sql`LOWER(${column}) = LOWER(${value})`;
}
/**
* Case-insensitive inequality for strings.
*/
function insensitiveNe(column, value) {
	return sql`LOWER(${column}) <> LOWER(${value})`;
}
//#endregion
//#region src/drizzle-adapter.ts
/**
* Derive the number of affected rows from a Drizzle write result.
*
* Drizzle returns the raw per-driver result for a non-returning write, so the
* count lives under a different field per driver: node-postgres / neon expose
* `rowCount`, postgres-js / bun-sql carry `count` on an Array subclass, mysql2
* reports `affectedRows` (in a result-header array), planetscale and other
* serverless drivers use `rowsAffected`, better-sqlite3 uses `changes`, and
* Cloudflare D1 nests the count under `meta.changes`. This normalizes them so
* write methods that depend on affected rows honor the adapter contract instead
* of leaking the raw driver result.
*/
function getAffectedRowCount(result, operation, context) {
	let count = 0;
	if (result && typeof result === "object" && "rowCount" in result) count = result.rowCount;
	else if (result && typeof result === "object" && typeof result.count === "number") count = result.count;
	else if (Array.isArray(result)) count = result.length > 0 && hasDriverRowCount(result[0]) ? readDriverRowCount(result[0]) : result.length;
	else if (hasDriverRowCount(result)) count = readDriverRowCount(result);
	if (typeof count !== "number") {
		logger.error(`[Drizzle Adapter] The result of the ${operation} operation is not a number. This is likely a bug in the adapter. Please report this issue to the Better Auth team.`, {
			result,
			...context
		});
		return 0;
	}
	return count;
}
function readDriverRowCount(result) {
	if (!result || typeof result !== "object") return void 0;
	if ("affectedRows" in result) return result.affectedRows;
	if ("rowsAffected" in result) return result.rowsAffected;
	if ("changes" in result) return result.changes;
	if ("meta" in result) {
		const meta = result.meta;
		if (meta && typeof meta === "object" && "changes" in meta) return meta.changes;
	}
}
function hasDriverRowCount(result) {
	return readDriverRowCount(result) !== void 0;
}
const drizzleAdapter = (db, config) => {
	let lazyOptions = null;
	let mysqlNoIdWarned = false;
	const createCustomAdapter = (db, inTransaction = false) => ({ getFieldName, getDefaultFieldName, getDefaultModelName, options, schema: baSchema }) => {
		if (config.provider === "mysql" && options.advanced?.database?.generateId === false && !mysqlNoIdWarned) {
			mysqlNoIdWarned = true;
			logger.warn("[Drizzle Adapter] MySQL does not support INSERT...RETURNING. With generateId set to false, the adapter uses best-effort fallback strategies (unique columns, full-field match) to retrieve inserted rows. For reliable behavior, use Better Auth's default ID generation, a custom generateId function, or generateId: \"serial\" for auto-increment.");
		}
		function getSchema(model) {
			const schema = config.schema || db._.fullSchema;
			if (!schema) throw new BetterAuthError("Drizzle adapter failed to initialize. Schema not found. Please provide a schema object in the adapter options object.");
			const schemaModel = schema[model];
			if (!schemaModel) throw new BetterAuthError(`[# Drizzle Adapter]: The model "${model}" was not found in the schema object. Please pass the schema directly to the adapter options.`);
			return schemaModel;
		}
		const withReturning = async (model, builder, data, where) => {
			if (config.provider !== "mysql") return (await builder.returning())[0];
			await builder.execute();
			const schemaModel = getSchema(model);
			const builderVal = builder.config?.values;
			if (where?.length) {
				const clause = convertWhereClause(where.map((w) => {
					if (data[w.field] !== void 0) return {
						...w,
						value: data[w.field]
					};
					return w;
				}), model);
				return (await db.select().from(schemaModel).where(...clause))[0];
			}
			const fetchInserted = async (tx) => {
				const builderId = builderVal?.[0]?.id?.value;
				if (builderId) return (await tx.select().from(schemaModel).where(eq(schemaModel.id, builderId)).limit(1).execute())[0] ?? null;
				if (data.id) return (await tx.select().from(schemaModel).where(eq(schemaModel.id, data.id)).limit(1).execute())[0] ?? null;
				if (options.advanced?.database?.generateId === "serial" && schemaModel.id) {
					const lastId = (await tx.select({ id: sql`LAST_INSERT_ID()` }).from(schemaModel).limit(1).execute())[0]?.id;
					if (lastId) return (await tx.select().from(schemaModel).where(eq(schemaModel.id, lastId)).limit(1).execute())[0] ?? null;
				}
				const modelSchema = baSchema[getDefaultModelName(model)]?.fields;
				if (modelSchema) for (const [fieldKey, fieldAttr] of Object.entries(modelSchema)) {
					if (!fieldAttr.unique) continue;
					const dbFieldName = getFieldName({
						model,
						field: fieldKey
					});
					const val = data[dbFieldName];
					if (val === void 0 || val === null) continue;
					if (!schemaModel[dbFieldName]) continue;
					const res = await tx.select().from(schemaModel).where(eq(schemaModel[dbFieldName], val)).limit(1).execute();
					if (res[0]) return res[0];
				}
				const conditions = [];
				for (const [key, val] of Object.entries(data)) {
					if (val === void 0 || !schemaModel[key]) continue;
					conditions.push(val === null ? isNull(schemaModel[key]) : eq(schemaModel[key], val));
				}
				if (conditions.length > 0) {
					const combined = and(...conditions);
					if (combined) {
						const res = await tx.select().from(schemaModel).where(combined).limit(2).execute();
						if (res.length === 1) return res[0];
					}
				}
				logger.warn(`[Drizzle Adapter] Unable to safely identify the inserted "${model}" row on MySQL. Enable Better Auth ID generation or use generateId: "serial" for reliable behavior.`);
				return null;
			};
			return inTransaction ? fetchInserted(db) : db.transaction(fetchInserted);
		};
		function convertWhereClause(where, model) {
			const schemaModel = getSchema(model);
			if (!where) return [];
			if (where.length === 1) {
				const w = where[0];
				if (!w) return [];
				const field = getFieldName({
					model,
					field: w.field
				});
				if (!schemaModel[field]) throw new BetterAuthError(`The field "${w.field}" does not exist in the schema for the model "${model}". Please update your schema.`);
				const isInsensitive = (w.mode ?? "sensitive") === "insensitive" && (typeof w.value === "string" || Array.isArray(w.value) && w.value.every((v) => typeof v === "string"));
				if (w.operator === "in") {
					if (!Array.isArray(w.value)) throw new BetterAuthError(`The value for the field "${w.field}" must be an array when using the "in" operator.`);
					if (isInsensitive) return [insensitiveInArray(schemaModel[field], w.value)];
					return [inArray(schemaModel[field], w.value)];
				}
				if (w.operator === "not_in") {
					if (!Array.isArray(w.value)) throw new BetterAuthError(`The value for the field "${w.field}" must be an array when using the "not_in" operator.`);
					if (isInsensitive) return [insensitiveNotInArray(schemaModel[field], w.value)];
					return [notInArray(schemaModel[field], w.value)];
				}
				if (w.operator === "contains") {
					if (isInsensitive && typeof w.value === "string") return [insensitiveIlike(schemaModel[field], `%${w.value}%`, config.provider)];
					return [like(schemaModel[field], `%${w.value}%`)];
				}
				if (w.operator === "starts_with") {
					if (isInsensitive && typeof w.value === "string") return [insensitiveIlike(schemaModel[field], `${w.value}%`, config.provider)];
					return [like(schemaModel[field], `${w.value}%`)];
				}
				if (w.operator === "ends_with") {
					if (isInsensitive && typeof w.value === "string") return [insensitiveIlike(schemaModel[field], `%${w.value}`, config.provider)];
					return [like(schemaModel[field], `%${w.value}`)];
				}
				if (w.operator === "lt") return [lt(schemaModel[field], w.value)];
				if (w.operator === "lte") return [lte(schemaModel[field], w.value)];
				if (w.operator === "ne") {
					if (w.value === null) return [isNotNull(schemaModel[field])];
					if (isInsensitive && typeof w.value === "string") return [insensitiveNe(schemaModel[field], w.value)];
					return [ne(schemaModel[field], w.value)];
				}
				if (w.operator === "gt") return [gt(schemaModel[field], w.value)];
				if (w.operator === "gte") return [gte(schemaModel[field], w.value)];
				if (w.value === null) return [isNull(schemaModel[field])];
				if (isInsensitive && typeof w.value === "string") return [insensitiveEq(schemaModel[field], w.value)];
				return [eq(schemaModel[field], w.value)];
			}
			const andGroup = where.filter((w) => w.connector === "AND" || !w.connector);
			const orGroup = where.filter((w) => w.connector === "OR");
			const andClause = and(...andGroup.map((w) => {
				const field = getFieldName({
					model,
					field: w.field
				});
				const isInsensitive = (w.mode ?? "sensitive") === "insensitive" && (typeof w.value === "string" || Array.isArray(w.value) && w.value.every((v) => typeof v === "string"));
				if (w.operator === "in") {
					if (!Array.isArray(w.value)) throw new BetterAuthError(`The value for the field "${w.field}" must be an array when using the "in" operator.`);
					if (isInsensitive) return insensitiveInArray(schemaModel[field], w.value);
					return inArray(schemaModel[field], w.value);
				}
				if (w.operator === "not_in") {
					if (!Array.isArray(w.value)) throw new BetterAuthError(`The value for the field "${w.field}" must be an array when using the "not_in" operator.`);
					if (isInsensitive) return insensitiveNotInArray(schemaModel[field], w.value);
					return notInArray(schemaModel[field], w.value);
				}
				if (w.operator === "contains") {
					if (isInsensitive && typeof w.value === "string") return insensitiveIlike(schemaModel[field], `%${w.value}%`, config.provider);
					return like(schemaModel[field], `%${w.value}%`);
				}
				if (w.operator === "starts_with") {
					if (isInsensitive && typeof w.value === "string") return insensitiveIlike(schemaModel[field], `${w.value}%`, config.provider);
					return like(schemaModel[field], `${w.value}%`);
				}
				if (w.operator === "ends_with") {
					if (isInsensitive && typeof w.value === "string") return insensitiveIlike(schemaModel[field], `%${w.value}`, config.provider);
					return like(schemaModel[field], `%${w.value}`);
				}
				if (w.operator === "lt") return lt(schemaModel[field], w.value);
				if (w.operator === "lte") return lte(schemaModel[field], w.value);
				if (w.operator === "gt") return gt(schemaModel[field], w.value);
				if (w.operator === "gte") return gte(schemaModel[field], w.value);
				if (w.operator === "ne") {
					if (w.value === null) return isNotNull(schemaModel[field]);
					if (isInsensitive && typeof w.value === "string") return insensitiveNe(schemaModel[field], w.value);
					return ne(schemaModel[field], w.value);
				}
				if (w.value === null) return isNull(schemaModel[field]);
				if (isInsensitive && typeof w.value === "string") return insensitiveEq(schemaModel[field], w.value);
				return eq(schemaModel[field], w.value);
			}));
			const orClause = or(...orGroup.map((w) => {
				const field = getFieldName({
					model,
					field: w.field
				});
				if (!schemaModel[field]) throw new BetterAuthError(`The field "${w.field}" does not exist in the schema for the model "${model}". Please update your schema.`);
				const isInsensitive = (w.mode ?? "sensitive") === "insensitive" && (typeof w.value === "string" || Array.isArray(w.value) && w.value.every((v) => typeof v === "string"));
				if (w.operator === "in") {
					if (!Array.isArray(w.value)) throw new BetterAuthError(`The value for the field "${w.field}" must be an array when using the "in" operator.`);
					if (isInsensitive) return insensitiveInArray(schemaModel[field], w.value);
					return inArray(schemaModel[field], w.value);
				}
				if (w.operator === "not_in") {
					if (!Array.isArray(w.value)) throw new BetterAuthError(`The value for the field "${w.field}" must be an array when using the "not_in" operator.`);
					if (isInsensitive) return insensitiveNotInArray(schemaModel[field], w.value);
					return notInArray(schemaModel[field], w.value);
				}
				if (w.operator === "contains") {
					if (isInsensitive && typeof w.value === "string") return insensitiveIlike(schemaModel[field], `%${w.value}%`, config.provider);
					return like(schemaModel[field], `%${w.value}%`);
				}
				if (w.operator === "starts_with") {
					if (isInsensitive && typeof w.value === "string") return insensitiveIlike(schemaModel[field], `${w.value}%`, config.provider);
					return like(schemaModel[field], `${w.value}%`);
				}
				if (w.operator === "ends_with") {
					if (isInsensitive && typeof w.value === "string") return insensitiveIlike(schemaModel[field], `%${w.value}`, config.provider);
					return like(schemaModel[field], `%${w.value}`);
				}
				if (w.operator === "lt") return lt(schemaModel[field], w.value);
				if (w.operator === "lte") return lte(schemaModel[field], w.value);
				if (w.operator === "gt") return gt(schemaModel[field], w.value);
				if (w.operator === "gte") return gte(schemaModel[field], w.value);
				if (w.operator === "ne") {
					if (w.value === null) return isNotNull(schemaModel[field]);
					if (isInsensitive && typeof w.value === "string") return insensitiveNe(schemaModel[field], w.value);
					return ne(schemaModel[field], w.value);
				}
				if (w.value === null) return isNull(schemaModel[field]);
				if (isInsensitive && typeof w.value === "string") return insensitiveEq(schemaModel[field], w.value);
				return eq(schemaModel[field], w.value);
			}));
			if (andGroup.length && orGroup.length) return [and(andClause, orClause)];
			if (andGroup.length) return [andClause];
			if (orGroup.length) return [orClause];
			return [];
		}
		function checkMissingFields(schema, model, values) {
			if (!schema) throw new BetterAuthError("Drizzle adapter failed to initialize. Drizzle Schema not found. Please provide a schema object in the adapter options object.");
			for (const key in values) {
				let fieldName;
				try {
					fieldName = getFieldName({
						model,
						field: key
					});
				} catch {
					fieldName = key;
				}
				if (!schema[fieldName]) throw new BetterAuthError(`The field "${key}" does not exist in the "${model}" Drizzle schema. Please update your drizzle schema or re-generate using "npx auth@latest generate".`);
			}
		}
		/**
		* Resolve the db.query key for a model.
		*
		* When `usePlural` is false (default), Better Auth uses singular model
		* names like "user", but Drizzle's db.query is keyed by the schema
		* export names (often plural like "users"). This function:
		*
		* 1. Tries the model name directly (works when schema keys match)
		* 2. If usePlural is set, tries appending "s"
		* 3. Falls back to scanning config.schema to find which db.query key
		*    corresponds to the same table object
		*/
		function getQueryModel(model) {
			if (db.query[model]) return model;
			if (config.usePlural) {
				const plural = `${model}s`;
				if (db.query[plural]) return plural;
			}
			if (config.schema) {
				const targetTable = config.schema[model];
				if (targetTable) {
					const fullSchema = db._.fullSchema;
					if (fullSchema) {
						for (const key of Object.keys(db.query)) if (fullSchema[key] === targetTable) return key;
					}
				}
			}
			return null;
		}
		return {
			async create({ model, data: values }) {
				const schemaModel = getSchema(model);
				checkMissingFields(schemaModel, model, values);
				return await withReturning(model, db.insert(schemaModel).values(values), values);
			},
			async findOne({ model, where, select, join }) {
				const schemaModel = getSchema(model);
				const clause = convertWhereClause(where, model);
				if (options.experimental?.joins) {
					const queryModel = getQueryModel(model);
					if (!db.query || !queryModel) {
						logger.error(`[# Drizzle Adapter]: The model "${model}" was not found in the query object. Please update your Drizzle schema to include relations or re-generate using "npx auth@latest generate".`);
						logger.info("Falling back to regular query");
					} else {
						let includes;
						const pluralJoinResults = [];
						if (join) {
							includes = {};
							const joinEntries = Object.entries(join);
							for (const [model, joinAttr] of joinEntries) {
								const limit = joinAttr.limit ?? options.advanced?.database?.defaultFindManyLimit ?? 100;
								const isUnique = joinAttr.relation === "one-to-one";
								const pluralSuffix = isUnique || config.usePlural ? "" : "s";
								includes[`${model}${pluralSuffix}`] = isUnique ? true : { limit };
								if (!isUnique) pluralJoinResults.push(`${model}${pluralSuffix}`);
							}
						}
						const res = await db.query[queryModel].findFirst({
							where: clause[0],
							columns: select?.length && select.length > 0 ? select.reduce((acc, field) => {
								acc[getFieldName({
									model,
									field
								})] = true;
								return acc;
							}, {}) : void 0,
							with: includes
						});
						if (res) for (const pluralJoinResult of pluralJoinResults) {
							const singularKey = !config.usePlural ? pluralJoinResult.slice(0, -1) : pluralJoinResult;
							res[singularKey] = res[pluralJoinResult];
							if (pluralJoinResult !== singularKey) delete res[pluralJoinResult];
						}
						return res;
					}
				}
				const res = await db.select(select?.length && select.length > 0 ? select.reduce((acc, field) => {
					const fieldName = getFieldName({
						model,
						field
					});
					return {
						...acc,
						[fieldName]: schemaModel[fieldName]
					};
				}, {}) : void 0).from(schemaModel).where(...clause);
				if (!res.length) return null;
				return res[0];
			},
			async findMany({ model, where, sortBy, limit, select, offset, join }) {
				const schemaModel = getSchema(model);
				const clause = where ? convertWhereClause(where, model) : [];
				const sortFn = sortBy?.direction === "desc" ? desc : asc;
				if (options.experimental?.joins) {
					const queryModel = getQueryModel(model);
					if (!queryModel) {
						logger.error(`[# Drizzle Adapter]: The model "${model}" was not found in the query object. Please update your Drizzle schema to include relations or re-generate using "npx auth@latest generate".`);
						logger.info("Falling back to regular query");
					} else {
						let includes;
						const pluralJoinResults = [];
						if (join) {
							includes = {};
							const joinEntries = Object.entries(join);
							for (const [model, joinAttr] of joinEntries) {
								const isUnique = joinAttr.relation === "one-to-one";
								const limit = joinAttr.limit ?? options.advanced?.database?.defaultFindManyLimit ?? 100;
								const pluralSuffix = isUnique || config.usePlural ? "" : "s";
								includes[`${model}${pluralSuffix}`] = isUnique ? true : { limit };
								if (!isUnique) pluralJoinResults.push(`${model}${pluralSuffix}`);
							}
						}
						let orderBy = void 0;
						if (sortBy?.field) orderBy = [sortFn(schemaModel[getFieldName({
							model,
							field: sortBy?.field
						})])];
						const res = await db.query[queryModel].findMany({
							where: clause[0],
							with: includes,
							columns: select?.length && select.length > 0 ? select.reduce((acc, field) => {
								acc[getFieldName({
									model,
									field
								})] = true;
								return acc;
							}, {}) : void 0,
							limit: limit ?? 100,
							offset: offset ?? 0,
							orderBy
						});
						if (res) for (const item of res) for (const pluralJoinResult of pluralJoinResults) {
							const singularKey = !config.usePlural ? pluralJoinResult.slice(0, -1) : pluralJoinResult;
							if (singularKey === pluralJoinResult) continue;
							item[singularKey] = item[pluralJoinResult];
							delete item[pluralJoinResult];
						}
						return res;
					}
				}
				let builder = db.select(select?.length && select.length > 0 ? select.reduce((acc, field) => {
					const fieldName = getFieldName({
						model,
						field
					});
					return {
						...acc,
						[fieldName]: schemaModel[fieldName]
					};
				}, {}) : void 0).from(schemaModel);
				const effectiveLimit = limit;
				const effectiveOffset = offset;
				if (typeof effectiveLimit !== "undefined") builder = builder.limit(effectiveLimit);
				if (typeof effectiveOffset !== "undefined") builder = builder.offset(effectiveOffset);
				if (sortBy?.field) builder = builder.orderBy(sortFn(schemaModel[getFieldName({
					model,
					field: sortBy?.field
				})]));
				return await builder.where(...clause);
			},
			async count({ model, where }) {
				const schemaModel = getSchema(model);
				const clause = where ? convertWhereClause(where, model) : [];
				return (await db.select({ count: count() }).from(schemaModel).where(...clause))[0].count;
			},
			async update({ model, where, update: values }) {
				const schemaModel = getSchema(model);
				const clause = convertWhereClause(where, model);
				return await withReturning(model, db.update(schemaModel).set(values).where(...clause), values, where);
			},
			async updateMany({ model, where, update: values }) {
				const schemaModel = getSchema(model);
				const clause = convertWhereClause(where, model);
				return getAffectedRowCount(await db.update(schemaModel).set(values).where(...clause), "updateMany", {
					model,
					where
				});
			},
			async delete({ model, where }) {
				const schemaModel = getSchema(model);
				const clause = convertWhereClause(where, model);
				return await db.delete(schemaModel).where(...clause);
			},
			async deleteMany({ model, where }) {
				const schemaModel = getSchema(model);
				const clause = convertWhereClause(where, model);
				return getAffectedRowCount(await db.delete(schemaModel).where(...clause), "deleteMany", {
					model,
					where
				});
			},
			async consumeOne({ model, where }) {
				const schemaModel = getSchema(model);
				const clause = convertWhereClause(where, model);
				const idField = getFieldName({
					model,
					field: "id"
				});
				const idColumn = schemaModel[idField];
				if (config.provider === "mysql") {
					const claimFromTransaction = async (tx) => {
						const target = (await tx.select().from(schemaModel).where(...clause).for("update").limit(1))[0];
						if (!target) return null;
						const targetId = target[idField] ?? target.id;
						if (targetId === void 0 || targetId === null || !idColumn) return null;
						return getAffectedRowCount(await tx.delete(schemaModel).where(eq(idColumn, targetId)).execute(), "consumeOne", {
							model,
							where
						}) > 0 ? target : null;
					};
					return inTransaction ? claimFromTransaction(db) : db.transaction(claimFromTransaction);
				}
				if (!idColumn) return null;
				const targetIds = db.select({ id: idColumn }).from(schemaModel).where(...clause).limit(1);
				return (await db.delete(schemaModel).where(inArray(idColumn, targetIds)).returning())[0] ?? null;
			},
			async incrementOne({ model, where, increment, set }) {
				const schemaModel = getSchema(model);
				const clause = convertWhereClause(where, model);
				const idField = getFieldName({
					model,
					field: "id"
				});
				const idColumn = schemaModel[idField];
				const assignments = {};
				for (const [field, delta] of Object.entries(increment)) {
					const columnName = getFieldName({
						model,
						field
					});
					const column = schemaModel[columnName];
					if (!column) throw new BetterAuthError(`The field "${field}" does not exist in the schema for the model "${model}". Please update your schema.`);
					assignments[columnName] = sql`${column} + ${delta}`;
				}
				if (set) for (const [field, value] of Object.entries(set)) {
					const columnName = getFieldName({
						model,
						field
					});
					if (!schemaModel[columnName]) throw new BetterAuthError(`The field "${field}" does not exist in the schema for the model "${model}". Please update your schema.`);
					assignments[columnName] = value;
				}
				if (config.provider === "mysql") {
					const mutateInTransaction = async (tx) => {
						const target = (await tx.select().from(schemaModel).where(...clause).for("update").limit(1))[0];
						if (!target) return null;
						const targetId = target[idField] ?? target.id;
						if (targetId === void 0 || targetId === null || !idColumn) return null;
						await tx.update(schemaModel).set(assignments).where(eq(idColumn, targetId)).execute();
						return (await tx.select().from(schemaModel).where(eq(idColumn, targetId)).limit(1).execute())[0] ?? null;
					};
					return inTransaction ? mutateInTransaction(db) : db.transaction(mutateInTransaction);
				}
				if (!idColumn) return null;
				const targetIds = db.select({ id: idColumn }).from(schemaModel).where(...clause).limit(1);
				return (await db.update(schemaModel).set(assignments).where(inArray(idColumn, targetIds)).returning())[0] ?? null;
			},
			options: config
		};
	};
	let adapterOptions = null;
	adapterOptions = {
		config: {
			adapterId: "drizzle",
			adapterName: "Drizzle Adapter",
			usePlural: config.usePlural ?? false,
			debugLogs: config.debugLogs ?? false,
			supportsUUIDs: config.provider === "pg" ? true : false,
			supportsJSON: config.provider === "pg" ? true : false,
			supportsArrays: config.provider === "pg" ? true : false,
			customTransformOutput: ({ data, fieldAttributes }) => {
				if (fieldAttributes.type === "date") {
					if (data === null || data === void 0) return data;
					return new Date(data);
				}
				return data;
			},
			transaction: config.transaction ?? false ? (cb) => db.transaction((tx) => {
				return cb(createAdapterFactory({
					config: {
						...adapterOptions.config,
						transaction: false
					},
					adapter: createCustomAdapter(tx, true)
				})(lazyOptions));
			}) : false
		},
		adapter: createCustomAdapter(db)
	};
	const adapter = createAdapterFactory(adapterOptions);
	return (options) => {
		lazyOptions = options;
		return adapter(options);
	};
};
//#endregion
export { drizzleAdapter };
